lib/recursive-methods.rb

```diff
@@ -1,49 +1,95 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+    raise ArgumentError if n < 0 
+    return 1 if n == 0
+    return n * factorial(n - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+    i = 0 
+    j = s.length - 1 
+    
+    while i < j 
+        temp = s[i] 
+        s[i] = s[j] 
+        s[j] = temp 
+        i += 1 
+        j -= 1 
+    end 
+    return s
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: O(n)
+# Space complexity: O(n)
+def reverse_inplace(s, i = 0, j = s.length - 1)
+    return s if i > j
+    temp = s[i]
+    s[i] = s[j] 
+    s[j] = temp
+    
+    return reverse_inplace(s, i + 1, j - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+    return 0 if n == 0
+    return 2 + bunny(n-1)
+end 
+
+# Time complexity: O(n)
+# Space complexity: O(n)
+def nested(s, i = 0, j = s.length - 1)
+    return true if i > j 
+    return false if (s[i] != "(") || (s[j] != ")") 
+    return nested(s, i + 1, j - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def nested(s)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: O(n)
+# Space complexity: O(n)
+def search(array, value, index = 0)
+    
+    return false if index > array.length
+    return true if array[index] == value
+
+    return search(array, value, index + 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def search(array, value)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: O(n)
+# Space complexity: O(n)
+def is_palindrome(s, i = 0, j = s.length - 1)
+    return true if i > j 
+    return false if s[i] != s[j]
+    return is_palindrome(s, i + 1, j - 1)
 end
 
 # Time complexity: ?
 # Space complexity: ?
-def is_palindrome(s)
-    raise NotImplementedError, "Method not implemented"
+### this one doesn't work yet, still can't figure it out :(
+def negative(n)
+    amount = n.to_s
+    return ("-" + amount).to_i
+end 
+
+def digit_match(n, m, count = 0, i = -1)
```

Comment:
Noted.  For a hint.  Remember that n % 10 will give you the ones digit and n / 10 will remove that digit.

So 1337 % 10 = 7
and 1337 / 10 = 133